cl/redisapp/leaderfollower/leaderfollower.go

```diff
@@ -0,0 +1,323 @@
+	lfm.leaderProc.Stop()
+	return false, nil
+}
+
+func (lfm *LeaderFollowerManager) followerWork(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			lfm.logger.Debug("Follower: Exiting")
+			return nil
+		default:
+			lfm.logger.Info("Follower: Performing follower tasks")
+
+			messages, err := lfm.readMessages(ctx)
+			if err != nil {
+				lfm.logger.Error("Follower: Error reading messages", "error", err)
+				continue
+			}
+
+			if len(messages) == 0 || len(messages[0].Messages) == 0 {
+				lfm.logger.Info("Follower: No messages to process")
+				return nil
+			}
+
+			// Process messages
+			for _, msg := range messages {
+				for _, field := range msg.Messages {
+					// Extract fields
+					payloadIDStr, ok := field.Values["payload_id"].(string)
+					executionPayloadStr, okPayload := field.Values["execution_payload"].(string)
+					senderInstanceID, okSenderID := field.Values["sender_instance_id"].(string)
+					if !ok || !okPayload || !okSenderID || payloadIDStr == "" || executionPayloadStr == "" || senderInstanceID == "" {
+						lfm.logger.Error("Follower: Invalid message format")
+						// Acknowledge the message to avoid reprocessing
+						if ackErr := lfm.stateManager.AckMessage(ctx, field.ID); ackErr != nil {
+							lfm.logger.Error("Follower: Failed to acknowledge message", "error", ackErr)
+						}
+						continue
+					}
+
+					// Ignore messages sent by self
+					if senderInstanceID == lfm.instanceID {
+						lfm.logger.Info("Follower: Ignoring own message", "PayloadID", payloadIDStr)
+						if ackErr := lfm.stateManager.AckMessage(ctx, field.ID); ackErr != nil {
+							lfm.logger.Error("Follower: Failed to acknowledge own message", "error", ackErr)
+						}
+						continue
+					}
+
+					lfm.logger.Info("Follower: Processing message", "PayloadID", payloadIDStr)
+
+					// Finalize block
+					if err := lfm.blockBuilder.FinalizeBlock(ctx, payloadIDStr, executionPayloadStr, field.ID); err != nil {
+						lfm.logger.Error("Follower: Failed to finalize block", "error", err)
+						continue
+					}
+
+					lfm.logger.Info("Follower: Successfully finalized block", "PayloadID", payloadIDStr)
+				}
+			}
+		}
+	}
+}
+
+func (lfm *LeaderFollowerManager) readMessages(ctx context.Context) ([]redis.XStream, error) {
+	// Try to read pending messages first
+	messages, err := lfm.stateManager.ReadMessagesFromStream(ctx, types.RedisMsgTypePending)
+	if err != nil {
+		lfm.logger.Error("Follower: Error reading pending messages", "error", err)
+		return nil, err
+	}
+
+	// If no pending messages, read new messages
+	if len(messages) == 0 || len(messages[0].Messages) == 0 {
+		messages, err = lfm.stateManager.ReadMessagesFromStream(ctx, types.RedisMsgTypeNew)
+		if err != nil {
+			lfm.logger.Error("Follower: Error reading new messages", "error", err)
+			return nil, err
+		}
+	}
+
+	return messages, nil
+}
+
+func (lfm *LeaderFollowerManager) Stop() error {
```

Comment:
Stop function doesnt actually stop anything. This is confusing. If user thinks he can call Stop and stop this, he would be blocked forever.

Secondly, it is better practice to do something like:
```
closed := make(chan struct{})
go func() {
  defer close(closed)
  wg.Wait()
}()

select {
  case <-time.After(5*time.Second):
    logger.Error("")
    return errors.New("workers still running")
  case <-closed:
    return nil
}
```

This way the wait is best effort and closing of the calling process is not blocked.